blocks/certifier.go

```diff
@@ -0,0 +1,413 @@
+			bid:        bid,
+			deadline:   now.Add(time.Duration(c.cfg.SignatureWaitDuration) * time.Second),
+			signatures: make([]types.CertifyMessage, 0, c.cfg.CertifyThreshold),
+		}
+	}
+}
+
+// CertifyIfEligible signs the hare output, along with its role proof as a certifier, and gossip the CertifyMessage
+// if the node is eligible to be a certifier.
+func (c *Certifier) CertifyIfEligible(ctx context.Context, logger log.Log, lid types.LayerID, bid types.BlockID) error {
+	// check if the node is eligible to certify the hare output
+	proof, err := c.oracle.Proof(ctx, lid, eligibility.CertifyRound)
+	if err != nil {
+		logger.With().Error("failed to get eligibility proof to certify", log.Err(err))
+		return err
+	}
+
+	eligibilityCount, err := c.oracle.CalcEligibility(ctx, lid, eligibility.CertifyRound, c.cfg.CommitteeSize, c.nodeID, proof)
+	if err != nil {
+		logger.With().Error("failed to check eligibility to certify", log.Err(err))
+		return err
+	}
+	if eligibilityCount == 0 { // not eligible
+		return nil
+	}
+
+	msg := types.CertifyMessage{
+		CertifyContent: types.CertifyContent{
+			LayerID:        lid,
+			BlockID:        bid,
+			EligibilityCnt: eligibilityCount,
+			Proof:          proof,
+		},
+	}
+	msg.Signature = c.signer.Sign(msg.Bytes())
+	data, err := codec.Encode(msg)
+	if err != nil {
+		logger.With().Panic("failed to serialize certify message", log.Err(err))
+		return err
+	}
+	if err = c.publisher.Publish(ctx, pubsub.BlockCertify, data); err != nil {
+		logger.With().Error("failed to send certify message", log.Err(err))
+		return err
+	}
+	return nil
+}
+
+// HandleCertifyMessage is the gossip receiver for certify message.
+func (c *Certifier) HandleCertifyMessage(ctx context.Context, _ p2p.Peer, msg []byte) pubsub.ValidationResult {
+	if c.isShuttingDown() {
+		return pubsub.ValidationIgnore
+	}
+
+	err := c.handleRawCertifyMsg(ctx, msg)
+	switch {
+	case err == nil:
+		return pubsub.ValidationAccept
+	case errors.Is(err, errMalformedData) == true:
+		return pubsub.ValidationReject
+	case errors.Is(err, errKnownCertMsg) == true || errors.Is(err, errMsgTooOld):
+		return pubsub.ValidationIgnore
+	default:
+		c.logger.WithContext(ctx).With().Error("failed to process certify message gossip", log.Err(err))
+		return pubsub.ValidationIgnore
+	}
+}
+
+// NumCached returns the number of layers being cached in memory.
+func (c *Certifier) NumCached() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return len(c.certMsgs)
+}
+
+func (c *Certifier) getCertState(logger log.Log, lid types.LayerID, bid types.BlockID) (certState, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.getCertStateLocked(logger, lid, bid)
+}
+
+func (c *Certifier) getCertStateLocked(logger log.Log, lid types.LayerID, bid types.BlockID) (certState, error) {
+	if ci, ok := c.certMsgs[lid]; !ok {
+		logger.Error("layer not registered for cert")
+		return unknown, errInternal
+	} else if ci.bid != bid {
+		logger.Error("block not registered for cert")
```

Comment:
some part of the committee can be malicious and will sign different block
we should collect certificates for all blocks, until one of the them has enough space units

Reply:
changed code to allow for multiple blocks.